deduction_rules.py

```diff
@@ -167,4 +167,162 @@ def line_square(sudoku):
+                    made_deduction |= _ban_numbers(sudoku,cell,deleted_numbers,"naked-trios",cells_used)
+        return made_deduction
+
+    return _apply_for_nines(search_and_ban_in_subset)
+
+def hidden_trios(sudoku):
+    """RULE: if only A,B,C cells (in one row/col/sec) has a number x,y,z then delete every number from A,B,C except x,y,z."""
+    def search_hidden_trios(elems):
+        res = dict()
+        where_can_go = {i:[] for i in range(1,10)} # i-th number in which indices of elems can go
+        for idx,elem in enumerate(elems):
+            for num in elem:
+                where_can_go[num].append(idx)
+        
+        for nums in combinations(range(1,10),3):
+            nums_can_go = tuple(set.union(set(where_can_go[nums[0]]), set(where_can_go[nums[1]]), set(where_can_go[nums[2]])))
+            if len(set(where_can_go[nums[0]])) > 1 and len(set(where_can_go[nums[1]])) > 1 and len(set(where_can_go[nums[2]])) > 1 and\
+                len(nums_can_go)==3:
+                res[nums_can_go] = nums
+        return res
+
+    def search_and_ban_in_subset(cells_to_check):
+        made_deduction = False
+        allowed_numbers = _allowed_numbers(sudoku,cells_to_check)
+        trios = search_hidden_trios(allowed_numbers)
+        for trio,except_nums in trios.items():
+            cell0 = cells_to_check[trio[0]]
+            cell1 = cells_to_check[trio[1]]
+            cell2 = cells_to_check[trio[2]]
+            print(cell0, cell1, cell2, except_nums)
+            cells_used = sudoku.allowed[cell0[0]][cell0[1]].notNones()+sudoku.allowed[cell1[0]][cell1[1]].notNones()+sudoku.allowed[cell2[0]][cell2[1]].notNones()
+            made_deduction |= _ban_numbers(sudoku, cell0, filter(lambda x: x not in except_nums, allowed_numbers[trio[0]]), "hidden-trio",cells_used)
+            made_deduction |= _ban_numbers(sudoku, cell1, filter(lambda x: x not in except_nums, allowed_numbers[trio[1]]), "hidden-trio",cells_used)
+            made_deduction |= _ban_numbers(sudoku, cell2, filter(lambda x: x not in except_nums, allowed_numbers[trio[2]]), "hidden-trio",cells_used)
+        return made_deduction
+
+    return _apply_for_nines(search_and_ban_in_subset)
+
+
+def yswing(sudoku):
+    '''RULE: if (i,j)=AC and (k,l)=BC and (i,l)=AB then remove C from (j,k).'''
+    # (i,j)=AC and (k,l)=BC
+    def allowed_nums_multicells(*args):
+        return set.intersection(*[set(_allowed_numbers(sudoku,[arg])[0]) for arg in args])
+    
+    def cells_in_same_sec(cell, col={0,1,2}, row={0,1,2}):
+        """Returns a list with the cells that are not set and are in the same sec as the input.
+        You can also set the (local)col or (local)row."""
+        sec = cell_section(cell[0],cell[1])
+        res = []
+        for i,j in product(range(3),range(3)):
+            if i in row and j in col:
+                new_cell = local_to_global(sec, i, j)
+                if len(allowed_nums_multicells(new_cell)) > 0:
```

Comment:
I'd create a separate function for `len(allowed_nums_multicells)` with a shorter name, but this is not absolutely necessary.

Reply:
Yes, but I am not so creative to find good short names.